substrate/frame/staking/src/tests.rs

```diff
@@ -4975,6 +4975,90 @@ fn on_finalize_weight_is_nonzero() {
 	})
 }
 
+mod sorted_list_provider_integration {
```

Comment:
would def. expect to see more tests here that check entire matrix of "staking states" x "staking operations"

staking states: unstaked, chilled, nominator, validator
staking actions: bond, increase bonded amount, reduce bonded amount, set validate, set nominate, chill, receive slash

And then make sure:

1. Correct events are sent for each item in the matrix.
2. Correct `VoterList::iter()` and `TargetList::iter()` are returned for each.